performant-software/dm-2
# Image layers

### What this PR does

- Per #288:
  - Adds support for image layers
  - Supports adding and deleting images in a layer stack with new toolbar icons
  - Supports naming layers with new toolbar "pencil" icon
  - Supports reordering layers with new toolbar "layer up" and "layer down" icons (#78)
  - Adds a drop down menu, up and down buttons for navigation (also includes #77)
- Per #76:
  - Retains zoom and position on layer change

### Status


### Steps to test

1. Visit https://dm-2-staging.herokuapp.com/
2. Log in
3. Open or create a project with Write or Admin permissions
4. Open or create an image document
5. Add a first image if one is not already present
6. "Check out" the document and click the <img width="18" alt="Add layer" src="https://user-images.githubusercontent.com/7234006/127913735-f1529267-26f4-448c-b01d-4b0674015d36.png"> icon in the toolbar to add a layer
7. Try adding layers by upload, web URL, and IIIF ([here is an example IIIF URL](https://stacks.stanford.edu/image/iiif/hg676jb4964%2F0380_796-44/info.json))
8. Verify that new layer names appear in toolbar, and that they are either the filename or "IIIF layer"
9. Verify that the new layer names appear in the layer navigation dropdown in the top left corner (may need to mouse over it to see it)
10. Click the pencil icon in the toolbar and rename a layer, then press the enter key or click the green checkmark to save the changes
11. Click the pencil icon in the toolbar and make some changes, then click the red X icon and verify that the changes are reverted
12. Use the dropdown as well as the ![down_grouphover](https://user-images.githubusercontent.com/7234006/127915152-11c359d0-cc66-4c26-a5dc-eb8c37f0ed65.png) and ![up_grouphover](https://user-images.githubusercontent.com/7234006/127915170-f1c5a12a-cab5-4889-ba88-bdf30284ef5e.png) icons to navigate between layers
   a. Verify that the up arrow is grayed out on the top layer, and that the down arrow is grayed out on the bottom layer
   b. Verify that zoom and position of the viewport are not affected by switching between layers
13. Hover over those icons to verify their tooltips reference "layers" and not "pages"
14. Use the <img width="17" alt="Move layer down" src="https://user-images.githubusercontent.com/7234006/127914970-1031ef47-dfb2-42b5-881c-04919f9c24a3.png"> and <img width="17" alt="Move layer up" src="https://user-images.githubusercontent.com/7234006/127914975-c0f4b559-c937-4218-ad59-67ca3e6728f3.png"> icons in the toolbar to reorder layers
15. Click the  <img width="18" alt="Delete layer" src="https://user-images.githubusercontent.com/7234006/127915355-946ee5b7-2d35-47d0-a5a4-b160da6a87b4.png"> icon and verify that a dialog comes up regarding layer deletion
16. Use that dialog to delete a layer and verify that it is deleted successfully
17. Delete all but one layer and verify that the layer navigation disappears, as well as the layer name in the toolbar 
18. Verify that highlights aren't affected by any of these actions
